### PR TITLE
Add torch.utils._content_store

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #100110
* #99834
* __->__ #99809

Implements a simple content-addressable store for storages (with tensors implemented as cheap references on top), enabling incremental serialization of tensors to disk, which I intend to use in the accuracy repro extractor.  Check the comment at the top of torch/utils/_content_store.py for more details on the intended use case.

One major piece of this PR is implementing the content hash for tensors.  For our prospective use case, we may need to repeatedly hash up to 80 GB of tensor data every time we snapshot (and we may snapshot multiple times).  Using a conventional cryptographic hash and hashing each snapshot would likely take on order of minutes, which seemed too slow to me.  So instead, I implemented a crappy hash function that can be run on GPU.  It is at least somewhat theoretically grounded: using random parameters generated by Philox, we use the standard shift-multiply and xor sum universal hash family.  The hash function is a bit dorky though; instead of properly doing 160-bit math, it just runs 32-bit hash five times and cats them together.  By the way, this sets the first precedent for kernel in PyTorch library which MUST be torch.compile'd to be run (in fact, this kernel does not run in eager mode because of the use of xor_sum, which doesn't actually exist in ATen.)

I had to add a few more primitives to inductor, namely randint (over the entire int range) and xor_sum.  Fortunately, these primitives are natively supported by Triton/C++, and so they were very easy to plumb through.  xor_sum is exposed as a prim, while randint special cases on when low/high span the entire 32-bit signed integer range.

Thanks to Jeff Johnson for letting me bounce ideas of him on a Saturday morning lol.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by ezyang** [comment]
Unfortunately the test case actually segfaults when run all in one go (but running the cpu or cuda separately passes?!)

**Comment by ezyang** [comment]
Hey @XiaobingSuper, when I patch in this PR and then run the test, I get a segfault in the generated C++ code:

```
extern "C" void kernel(const long* seed0,
                       const int* in_ptr1,
                       int* out_ptr0,
                       const long ks0)
{
    {
        {
            int tmp_acc0 = 0;
            for(long i0=static_cast<long>(0L); i0<static_cast<long>(ks0); i0+=static_cast<long>(1L))
            {
                auto tmp0 = seed0[static_cast<long>(0L)];
                auto tmp13 = in_ptr1[static_cast<long>(i0)];
                auto tmp1 = static_cast<int>(65535);
                auto tmp2 = decltype(tmp0)(tmp0 ^ tmp1);
                auto tmp3 = static_cast<int>(i0);
                auto tmp4 = static_cast<int>(randint_cpu(tmp2, tmp3));;
                auto tmp5 = std::abs(tmp4);
                auto tmp6 = static_cast<int>(2147483647);
```

specifically `in_ptr1` is invalid. What's a good way to debug this?

**Comment by ezyang** [comment]
I guess there's a guard problem, because the pointer is a CUDA pointer

**Comment by ezyang** [comment]
Found the bug

File: benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_dynamic_training.csv [hunk 1]
```diff
@@ -1,7 +1,7 @@
 name,accuracy,graph_breaks
 AlbertForMaskedLM,fail_to_run,4
 AlbertForQuestionAnswering,fail_to_run,4
-AllenaiLongformerBase,pass,159
+AllenaiLongformerBase,pass,160
```
Comment (anijain2305) [inline benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_dynamic_training.csv:4]: Why are these happening?

File: torch/utils/_content_store.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+import torch
+import torch._prims as prims
+import torch._utils
+import torch.nn.functional as F
+
+from torch.multiprocessing.reductions import StorageWeakRef
+
+
+def hash_storage(storage):
```
Comment (voznesenskym) [inline torch/utils/_content_store.py:42]: types please, here and everywhere in this file

File: torch/utils/_content_store.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    #   storages/
+    #     00/
+    #       0000..00
+    #   tensors/
+    #     name
+    def __init__(self, loc):
+        self.loc = loc
+        self.seen_storage_hashes = set()
+
+    def write_storage(self, storage):
```
Comment (voznesenskym) [inline torch/utils/_content_store.py:101]: prefix `_`? Seems like write_tensor is the API, not write_storage?

File: torch/utils/_content_store.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import torch._utils
+import torch.nn.functional as F
+
+from torch.multiprocessing.reductions import StorageWeakRef
+
+
+def hash_storage(storage):
+    # Use of torch.compile is mandatory for (1) good memory usage
+    # and (2) xor_sum implementation
+    @torch.compile(dynamic=True)
```
Comment (voznesenskym) [inline torch/utils/_content_store.py:45]: Pretty cool. We should leave a TODO here to AOT this once and store it, theres no need to recompile it every time, right?

File: torch/utils/_content_store.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        # don't use this for stable hashing.
+        #
+        # This assumes fixed length so you're also obligated to bucket
+        # by the length of tensor as well
+        return prims.xor_sum((a * x + b).int(), [0])
+
+    with torch.random.fork_rng(
+        [storage.device] if storage.device.type != "cpu" else []
+    ):
+        torch.manual_seed(0)  # this can be anything, just needs to be fixed
```
Comment (voznesenskym) [inline torch/utils/_content_store.py:73]: Will this change manual seed for the rest of the program? Do you need to set it back?

File: torch/utils/_content_store.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    def write_storage(self, storage):
+        h = hash_storage(storage)
+        if h in self.seen_storage_hashes:
+            return h
+        # TODO: consider not using torch.save for this; we don't actually
+        # need any metadata for the storage
+        subfolder = os.path.join(self.loc, "storages", h[:2])
+        os.makedirs(subfolder, exist_ok=True)
+        torch.save(storage, os.path.join(subfolder, h))
+        self.seen_storage_hashes.add(h)
```
Comment (voznesenskym) [inline torch/utils/_content_store.py:110]: nit: If you write a helper, you can use @lru_cache for this and not need to manage your own structure

File: torch/utils/_content_store.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        s = torch.load(
+            os.path.join(self.loc, "storages", h[:2], h), weights_only=True
+        )._untyped_storage
+        self.storage_cache[h] = StorageWeakRef(s)
+        return s
+
+    def read_tensor(self, name):
+        dtype, h, storage_offset, size, stride, metadata = torch.load(
+            os.path.join(self.loc, "tensors", name), weights_only=True
+        )
```
Comment (voznesenskym) [inline torch/utils/_content_store.py:151]: What if name is not there? does this rase under torch.load?

**Review by voznesenskym — APPROVED** [review]
Lots of questions that can be resolved w/ docs and types

File: benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_dynamic_training.csv (hunk 1 above)
Comment (voznesenskym) [reply benchmarks/dynamo/ci_expected_accuracy/inductor_huggingface_dynamic_training.csv:4]: > Do we know why are these happening?

https://github.com/pytorch/pytorch/pull/99823

Subsumed by this PR, but pulled into the above.

File: torch/testing/_internal/common_cuda.py [hunk 8]
```diff
@@ -20,6 +20,7 @@
 
 SM53OrLater = LazyVal(lambda: torch.cuda.is_available() and torch.cuda.get_device_capability() >= (5, 3))
 SM60OrLater = LazyVal(lambda: torch.cuda.is_available() and torch.cuda.get_device_capability() >= (6, 0))
+SM70OrLater = LazyVal(lambda: torch.cuda.is_available() and torch.cuda.get_device_capability() >= (8, 0))
```
Comment (bdhirsh) [inline torch/testing/_internal/common_cuda.py:23]: should this be `and torch.cuda.get_device_capability() >= (7, 0))`? Also, is this used anywhere?

File: torch/utils/_content_store.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+        subfolder = os.path.join(self.loc, "storages", h[:2])
+        os.makedirs(subfolder, exist_ok=True)
+        torch.save(storage, os.path.join(subfolder, h))
+        self.seen_storage_hashes.add(h)
+        return h
+
+    def write_tensor(self, name, t):
+        storage = t.untyped_storage()
+        h = self.write_storage(storage)
+        # TODO: Support more advanced snapshotting of requires_grad/grad/etc
```
Comment (bdhirsh) [inline torch/utils/_content_store.py:137]: My first thought is that this would be pretty important for any kind of minifying. But I guess for below-aot-autograd minifying, this isn't really an issue, since the graph you're snapshotting will be a post-autograd graph? (But for minifying a dynamo graph, this will be necessary for minifying any type of training graph).

File: torch/utils/_content_store.py (hunk 3 above)
Comment (ezyang) [reply torch/utils/_content_store.py:122]: I actually use write_storage directly in the next PR, they're both valid APIs.

File: torch/utils/_content_store.py (hunk 6 above)
Comment (ezyang) [reply torch/utils/_content_store.py:131]: Can't easily use lru_cache as storage isn't hashable. And it seems dorky to pass the StorageWeakRef as the input to the function?

File: torch/utils/_content_store.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+        torch.manual_seed(0)  # this can be anything, just needs to be fixed
+        x = torch.empty(0, dtype=torch.uint8, device=storage.device).set_(storage)
+        # The dtype-casting view cannot be compiled, and so the
+        # padding/reshaping also needs to be done externally even
+        # though it could be profitably fused
+        pad = x.numel() % 4
+        if pad > 0:
+            x = F.pad(x, (0, pad), "constant", 0)
+        x = x.view(torch.int32)
+        torch._dynamo.mark_dynamic(x, 0)
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/utils/_content_store.py:103]: Well, I mean, it's not actually guaranteed to be fully dynamic, but this is essentially a pointwise-reduction on a 1D tensor, so yeah, we do expect a single kernel to work here.

File: torch/utils/_content_store.py (hunk 9 above)
Comment (ezyang) [reply torch/utils/_content_store.py:137]: Yeah, exactly. Pre-aot we absolutely need this, and as I've been trying to use the new accuracy repro extractor on some real world bugs, I've noticed that pre-aot may be more important than post-aot. Another PR...

File: torch/utils/_content_store.py (hunk 5 above)
Comment (ezyang) [reply torch/utils/_content_store.py:94]: OK so apparently the random api is full of footguns. Rewrite is much better.

File: torch/testing/_internal/common_cuda.py (hunk 8 above)
Comment (ezyang) [reply torch/testing/_internal/common_cuda.py:23]: It used to be used to gate torch.compile tests, but then I made it fallback to SHA-1 so that skip could be removed. Still, it seems useful to have.
